### PR TITLE
Reduce the number of captured name resolution environments in unions and records

### PR DESCRIPTION
**Pull request by kerams**
Instead of incrementally inserting an environment for each field/case, we create only one for the entire representation scope as these entities cannot refer to each other.

Also fixes https://github.com/dotnet/fsharp/issues/13076.

Before

devenv_Tw2UxfsBOj

After

devenv_pyvfqBRRXZ

The difference in `SyntaxTree.fs` is 708 vs 1,054.

### CODE REVIEW

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 1]
```diff
@@ -3278,25 +3285,17 @@ module EstablishTypeDefinitionCores =
...
             let writeFakeUnionCtorsToSink (unionCases: UnionCase list) = 
                 let nenv = envinner.NameEnv
                 // Constructors should be visible from IntelliSense, so add fake names for them 
                 for unionCase in unionCases do
                     let info = UnionCaseInfo(thisTyInst, mkUnionCaseRef thisTyconRef unionCase.Id.idText)
-                    let nenv' = AddFakeNameToNameEnv unionCase.Id.idText nenv (Item.UnionCase(info, false)) 
-                    // Report to both - as in previous function
                     let item = Item.UnionCase(info, false)
                     CallNameResolutionSink cenv.tcSink (unionCase.Range, nenv, item, emptyTyparInst, ItemOccurence.Binding, ad)
-                    CallEnvSink cenv.tcSink (unionCase.Id.idRange, nenv', ad)
```
Comment (kerams) [inline src/Compiler/Checking/CheckDeclarations.fs:3299]: Contrary to these comments, we don't need the values for Intellisense. Perhaps completions relied on them historically.

**Comment by kerams** [comment]
Can't run the failing test locally because I'm getting `TypeInitializationException` since the .NET 8 SDK switch.

Doesn't this look right though?

image

**Comment by edgarfgp** [comment]
> Can't run the failing test locally because I'm getting `TypeInitializationException` since the .NET 8 SDK switch.
> 
> Doesn't this look right though?
> 
> image


Had a similar issues and I fixed by git cleaning it and then only build the FSharp.Compiler.Service.sln alone without using the build.cmd part . Hope it helps

File: tests/FSharp.Compiler.Service.Tests/TooltipTests.fs [hunk 2]
```diff
@@ -189,7 +189,7 @@ type Bar = {
 }
 """
 
-    testXmlDocFallbackToSigFileWhileInImplFile sigSource implSource 5 9 "    SomeField: int" [ "SomeField" ] "Some sig comment on record field"
+    testXmlDocFallbackToSigFileWhileInImplFile sigSource implSource 5 13 "    SomeField: int" [ "SomeField" ] "Some sig comment on record field"
```
Comment (kerams) [inline tests/FSharp.Compiler.Service.Tests/TooltipTests.fs:192]: We need to search with the end position of the identifier, which is what VS does.

**Review by T-Gro — APPROVED** [review]

**Comment by vzarytovskii** [comment]
> Can't run the failing test locally because I'm getting `TypeInitializationException` since the .NET 8 SDK switch.
> 
> Doesn't this look right though?
> 
> image
> 

Uh. That doesn't sound right. Do you have the specific issue?

**Review by 0101 — APPROVED** [review]

**Review by edgarfgp — APPROVED** [review]

**Review by vzarytovskii — APPROVED** [review]

**Pull request merged**